lnrpc/invoicesrpc/addinvoice_test.go

```diff
@@ -0,0 +1,586 @@
+package invoicesrpc
+
+import (
+	"encoding/hex"
+	"errors"
+	"math/big"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/zpay32"
+	"github.com/stretchr/testify/mock"
+)
+
+type hopHintsConfigMock struct {
+	mock.Mock
+}
+
+// IsChannelActive mocks channel active checks.
+func (h *hopHintsConfigMock) IsChannelActive(chanID lnwire.ChannelID) bool {
+	args := h.Mock.Called(chanID)
+	return args.Bool(0)
+}
+
+// IsPublicNode mocks node public state lookup.
+func (h *hopHintsConfigMock) IsPublicNode(pubKey [33]byte) (bool, error) {
+	args := h.Mock.Called(pubKey)
+	return args.Bool(0), args.Error(1)
+}
+
+// FetchChannelEdgesByID mocks channel edge lookup.
+func (h *hopHintsConfigMock) FetchChannelEdgesByID(chanID uint64) (
+	*channeldb.ChannelEdgeInfo, *channeldb.ChannelEdgePolicy,
+	*channeldb.ChannelEdgePolicy, error) {
+
+	args := h.Mock.Called(chanID)
+
+	// If our error is non-nil, we expect nil responses otherwise. Our
+	// casts below will fail with nil values, so we check our error and
+	// return early on failure first.
+	err := args.Error(3)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	edgeInfo := args.Get(0).(*channeldb.ChannelEdgeInfo)
+	policy1 := args.Get(1).(*channeldb.ChannelEdgePolicy)
+	policy2 := args.Get(2).(*channeldb.ChannelEdgePolicy)
+
+	return edgeInfo, policy1, policy2, err
+}
+
+// TestSelectHopHints tests selection of hop hints for a node with private
+// channels.
+func TestSelectHopHints(t *testing.T) {
```

Comment:
Any suggestion on how to make this kind of test more reviewable? I often struggle to get through massive amount of test data + cases. Perhaps breaking test cases up into different commits?

Reply:
I think these tests are actually very nice and readable once the reviewer understands the whole `SelectHopHint` logic deeply. Adding more commits may add a rhythm to the review that could be helpful but sum of cycles would probably be the same :)